packages/actor-rdf-resolve-quad-pattern-sparql-json/lib/ActorRdfResolveQuadPatternSparqlJson.ts

```diff
@@ -26,6 +26,51 @@ export class ActorRdfResolveQuadPatternSparqlJson
     super(args);
   }
 
+  /**
+   * Replace all blank nodes in a pattern with variables.
+   * If the pattern contains no blank nodes the original pattern gets returned.
+   * @param {RDF.Quad} pattern A quad pattern.
+   * @return {RDF.Quad} A quad pattern with no blank nodes.
+   */
+  public static replaceBlankNodes(pattern: RDF.Quad): RDF.Quad {
+    const variableNames: string[] = getVariables(getTerms(pattern)).map((v) => v.value);
+    const result: { [id: string]: RDF.Term } = {};
+    // track the names the blank nodes get mapped to (required if the name has to change)
+    const blankMap: { [id: string]: string } = {};
+    let changed = false;
+
+    // for every position, convert to a variable if there is a blank node
+    for (const pos of QUAD_TERM_NAMES) {
+      const term: RDF.Term = (<any> pattern)[pos];
+      if (term.termType === 'BlankNode') {
+        let name = term.value;
+        if (blankMap[name]) {
+          name = blankMap[name];
+        } else {
+          if (variableNames.indexOf(name) >= 0) {
```

Comment:
Using a hash instead of an array here would make this slightly faster.

Reply:
I'm actually wondering know if there might be a reason why the array could be faster (since it can contain at most 2 values) :D

Reply:
Oh, you're right. Probably doesn't matter then.